### ISSUE TITLE
Bad diagnostic for `any P!`

### ISSUE THREAD

**Issue by AnthonyLatsis** [apple/swift#72662]
### Reproduction

```swift
// error: using '!' is not allowed here; perhaps '?' was intended? [implicitly_unwrapped_optional_in_illegal_position]
// error: optional 'any' type must be written '(any P2)?' [incorrect_optional_any]
let _: any P!
```

### Expected behavior

A single error, `optional 'any' type must be written '(any P)!'`, as in the case of `let _: any P?`.

### Environment

Swift version 6.0-dev (LLVM 48dc0c6fa01ce9e, Swift 19fbe5b33d423d1)
Target: x86_64-apple-macosx14.0

### Additional information

This https://github.com/apple/swift/labels/good%20first%20issue will suit an experienced Swift user.

**Comment by CrazyFanFan** [comment]
@AnthonyLatsis  Would modifying `PrintOptions::forDiagnosticArguments()` be the correct solution to this problem? If so, I believe I could attempt to resolve it.

**Comment by AnthonyLatsis** [comment]
No, this is a problem with the diagnosis of invalid code, not the format used to print diagnostic messages.

**Comment by fabricerenard12** [comment]
Hi, if no one is already addressing this issue I would love to start working on it.

**Comment by iMostfa** [comment]
Hi @AnthonyLatsis i'm interest in this issue, 

i've made a PoC solution for this issue here, i'm not sure if it's the right way to fix it though, 71d3c4f1a6730d50d05219223216dcdde9f07ed4

i wanted to avoid touching `TypeResolver::resolveImplicitlyUnwrappedOptionalType`
and instead try to fix it from `TypeResolver::resolveExistentialType`

but i found that it's not possible as already too late we get a diagnostics emitted from: `TypeResolver::resolveImplicitlyUnwrappedOptionalType` before we emit the another (correct diagnostics in this case) from `resolveExistentialType`:



i ran tests in `sema` and it looks like my fix doesn't cause any regressions, what do you think ?

this's the current state after applying my fix: 

<img width="1257" alt="Screenshot 2024-06-16 at 18 11 46" src="https://github.com/apple/swift/assets/2325884/45e458e0-4d3f-4c7b-95d3-485b150f7153">


and tests in `sema` are ✅

would apperciate if i get this one assigned to me, and a hint if my approach needs to be changed

**Comment by AnthonyLatsis** [comment]
Hi, Mostfa! We prefer doing code review through pull requests, so you are very much encouraged to open one to `apple:main` if you are ready to discuss a solution.

**Comment by iMostfa** [comment]
Thanks @AnthonyLatsis For the quick response 

PR Opened https://github.com/apple/swift/pull/74461
